### PR TITLE
android: attempting to respect Android's cleartext policies

### PR DESCRIPTION
**Pull request by alyssawilk**
Rejecting requests which don't have an https scheme if Android disallows cleartext

Risk Level: High 
Testing: NEED HELP HERE
Docs Changes: n/a
Release Notes: inline
Fixes https://github.com/envoyproxy/envoy-mobile/issues/1572

### CODE REVIEW

File: library/common/jni/BUILD [hunk 1]
```diff
@@ -199,3 +181,24 @@ cc_library(
...
+        "jni_support.h",
+        "jni_utility.h",
+        "jni_version.h",
+    ],
+    deps = [
+        "//library/common/jni/import:jni_import_lib",
+        "//library/common/types:c_types_lib",
+        "@envoy//source/common/common:assert_lib",
+    ] + select({
+        "@envoy//bazel:android_logger": [":ndk_jni_support"],
```
Comment (alyssawilk) [inline library/common/jni/BUILD:201]: @keith this is probably not the right select for using android - do you know if there's a better opton?

Comment (keith) [reply library/common/jni/BUILD:201]: We have this internally https://github.com/envoyproxy/envoy/pull/21459 seems useful upstream at this point, wdyt

File: library/common/http/client.cc [hunk 2]
```diff
@@ -479,6 +480,16 @@ void Client::sendHeaders(envoy_stream_t stream, envoy_headers headers, bool end_
   if (direct_stream) {
     ScopeTrackerScopeState scope(direct_stream.get(), scopeTracker());
     RequestHeaderMapPtr internal_headers = Utility::toRequestHeaders(headers);
+
+    // This is largely a check for the android platform: is_cleartext_permitted
+    // is a no-op for other platforms.
+    if (internal_headers->getSchemeValue() != "https" &&
+        !is_cleartext_permitted(internal_headers->getHostValue())) {
+      direct_stream->request_decoder_->sendLocalReply(
+          Http::Code::BadRequest, "Cleartext is not permitted", nullptr, absl::nullopt, "");
```
Comment (goaway) [inline library/common/http/client.cc:489]: Rather than issuing the direct reply, could we simply call the on_error callback here?

**Comment by goaway** [comment]
Thanks @alyssawilk! Overall, I think the setup here looks good, but let me know what you think about the comment above.

File: library/common/http/client.cc (hunk 2 above)
Comment (goaway) [reply library/common/http/client.cc:489]: I guess the current setup will result in the filter chain being created, whereas firing the callback directly wouldn't(?). I don't know if that's desirable or not. Maybe it is, if someone is listening for errors in a filter?

But I suppose I'm fine either way, as long as we're consistent about we handle this sort of thing for any other cases moving forward.

**Review by goaway — APPROVED** [review]
Thanks @alyssawilk!

Like we talked about previously, I will work on an integration test for this once it's merged.

**Pull request merged**